control-plane/agents/src/bin/core/controller/scheduling/volume.rs

```diff
@@ -448,6 +480,24 @@ impl ReplicaRemovalCandidates {
             None
         }
     }
+    fn ag_restricted_nodes(&self) -> Option<&HashSet<NodeId>> {
+        let ag = self.context.affinity_group.as_ref()?;
+        if ag.restricted_nodes.is_empty() {
+            return None;
+        }
+        //     N
+        // R 1 2 3
+        // 1 x x
+        // 2 x x
+        // 3 x x x
+        // todo: In this case, shouldn't we block removal or r3n3 ?
```

Comment:
Indeed! But then how do we do it? Non single replicas also in ag_restricted_nodes? But that might not work well

Reply:
Possibly, though it's probably still fine as long as the user doesn't intend to reduce replicas down to 1.
This needs some more consideration, I'll leave it out of this PR since it can already happen anyway.